### ISSUE TITLE
Let modules specify dependencies

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/dependencies#1]
The first task in making this work is letting modules specify dependencies. 

There is a [plugin dependencies manager](https://github.com/xwp/wp-plugin-dependencies/), and it lets plugins specify other plugins in the `Depends:` file header. This is just a comma separated list of plugin names. While this is a very simple approach, it may be too simple for our purposes.

There are several things that we need to keep in mind. First, we will need to be able to specify more than just other modules as dependencies. In fact, at present the most important thing is to be able to specify which plugins are required for a module to work. And we can't rule out the possibility that in future we'll want to be able to specify themes as dependencies as well.

The next thing to consider is that we aren't necessarily requiring all dependencies from the same repository. Some plugins may be from WordPress.org, while others may be premium editions or plugins from other sites, like CodeCanyon. Because of this, we'll likely need to be able to specify exactly where this dependency is intended to come from.

Another important thing to consider which the plugin dependency plugin doesn't handle at all is the issue of version requirements.

Providing all of these features in WordPress file headers is going to be not ideal at best, and horrible at worst (most likely). This leads us to search for a better way.

Enter composer. Composer-like configuration provided via a JSON file fulfills all of our needs, and paves the way for us to add many more features in the future, like "recommended" dependencies. To be clear, I'm not talking about using composer in any way, just borrowing the concept of a JSON config file for dependencies. I think that this is the way to do it.

The only downside I see at the moment is that it is impossible to specify dependencies for a single-file module. Of course, single-file modules aren't really a concern, because modules are always placed in their own directory when installed through the administration screens.

**Comment by JDGrimes** [comment]
Another thing to consider is that modules may need to specify different types of dependencies. For example:
- `requires` - Things that are required for the module to work. Without them, you will experience problems, maybe even the white screen of death.
- `recommended` - Things that are recommended to use with the module. The module will work without them, but some features may be crippled, but the site won't break entirely.
- `incompatible` - Things with known incompatibilities. If the the module is used along with them it will result in problems or even break the site.
- `not recommended` - Things with known incompatibilities. If the module is used along with them some features may be crippled, but things won't blow up.
- `integrated` - Things that aren't required, or necessarily needed for all sites, but which the module will integrate with if present. 

There might be other useful types as well.

Also, each of these can also apply to dependency versions. For example:

``` json
{
  "wordpress": {
    "recomended": "3.8+",
    "required": "3.7+"
  } 
}
```

We need to decide the general structure of the JSON object, and we also need to decide what the syntax for versions is.

**Comment by JDGrimes** [comment]
We'd probably do something similar to [composer](https://getcomposer.org/doc/01-basic-usage.md#package-versions), but possibly not as complex.

**Comment by JDGrimes** [comment]
> We need to decide the general structure of the JSON object

To that end, let's consider how we'll actually be using this data. 
- When a module is activated, we'll want to see what requirements it has.
- When an entity is deactivated, we'll want to see if it is required by anything.
- When an entity is updated, we'll want to check for incompatibilities with the new version.

For all of these, we'll want to get the entities listed by dependency type, so we can easily get all the `required` deps.

Actually, now that I think about it, I'm not sure that we need to worry so much about the syntax of the files. Maybe we'll want to parse them all each time anyway and we could arrange the data into whatever kind of lookup tree we want. However, that could be a time consuming process, so probably not.

So I'm thinking that we'll want to make the syntax something like this:

``` json
{
    "required": {
        "wordpress": {
             "recommended": "3.9+",
             "required": "3.7+"
        },
        "plugins": { ... }
    }
}
```

In other words: `degree of dependency -> type of entity -> entity -> degree of dependency -> version`

WordPress would be a special case, with the `entity ->` level omitted.

**Comment by JDGrimes** [comment]
Something else that we'd want to do is check if an update for a module has more dependencies that aren't met. But to do that we shouldn't have to download the update. We should just be able to get the information over an API. I just happened to think whether this should all be done over API with no `wordpoints-dependency.json` file at all. But then it wouldn't work if the API was down, which would be sad.

**Comment by JDGrimes** [comment]
Really, we want to go beyond modules specifying dependencies, and let all the things specify dependencies this way. Maybe we should develop this a library used by WordPoints which could maybe even get merged into WordPress eventually.

**Comment by JDGrimes** [comment]
I just found another plugin dependency manager: http://tgmpluginactivation.com/ It might be useful, since it does more of the things we need, like non-wordpress.org plugins.

**Comment by JDGrimes** [comment]
A good read on plugin dependencies that proposes something similar: http://ttmm.io/tech/wordpress-dependency-management/

Several things to consider here, that are more about implementation:
- What should be hidden from the user, and what should be transparent?
- What should be done with dependency conflicts?

**Comment by JDGrimes** [comment]
Related WordPress tickets: [WP#10190](https://core.trac.wordpress.org/ticket/10190), [WP#11308](https://core.trac.wordpress.org/ticket/11308), [WP#12612](https://core.trac.wordpress.org/ticket/12612), [WP#13296](https://core.trac.wordpress.org/ticket/13296).

Other related stuff:
- [Plugin Dependency for WordPress Themes](http://www.onequarterenglish.co.uk/plugin-dependency-for-wordpress-themes)
- [WordPress Plugin Dependency Programming ](http://www.cleancode.co.nz/blog/481/wordpress-plugin-dependency-programming)
- [WordPress needs plugin dependencies](http://lucdebrouwer.nl/wordpress-needs-plugin-dependencies/)
- [A Case For Dependency Management with WordPress](https://tommcfarlin.com/dependency-management-with-wordpress/)

After reading everything that I can get my hands on with regard to WordPress and plugin dependencies, I think that this is going to require some careful thought. 

We need to:
- Identify a problem(s) we are trying to solve.
- Address concerns brought up in the tickets above as to why the idea was rejected.
- Research other dependency systems, like Drupal's.

**Comment by JDGrimes** [comment]
Problem 1: User installs a module but the plugin(s) it integrates with aren't installed.

Why this is a problem: Users _will_ actually do this. When the user activates the module, nothing will happen (ideally—in a worst case scenario they get the white screen of death).

Minimum Solution: The module shows a message when it is activated telling the user they need to install/activate its dependent plugin(s) for it to work.

WWJD Solution: The dependencies are automatically installed and activated.

Real Life Solution: We should detect that the dependencies aren't installed, and when the module is activated give the user the option to install and activate them automatically.

Notes:
- We want to be transparent about this, because the user is installing something that is visible. That is, the dependency has forward-facing features, or, at the very least, will be visible in the list of installed plugins. If the user sees these features (or the plugin row) and doesn't know where they have come from, this will cause confusion.

---

Problem 2: User deactivates a plugin that a module depends on.

Why this is a problem: The user may not have realized that the module had this plugin as a dependency, and may wonder why it has stopped working (worst case scenario they will suddenly get the WSOD).

Minimum Solution: The module shows a message to the user when it detects a missing dependency.

WWJD Solution: The module is automatically deactivated, or the plugin remains active and they are told why, depending on the user's intention.

Real Life Solution: The fact that the plugin is a dependency is displayed to the user, and they don't have the option to deactivate it alone, they have too choose to deactivate both the plugin and the module.

---

Problem 3: User updates a plugin to a version not compatible with the module.

Why this is a problem: The module may stop working, or be crippled (or they might get the WSOD).

Minimum Solution: This is actually non-trivial for the module itself to handle. It probably won't know about the incompatibilities in most cases. It can use `*_exists()` calls to avoid total disaster, though.

WWJD Solution: The incompatibilities are instantaneous resolved.

Real Life Solution: There are two parts to this. We first need to be able to determine whether there are incompatibilities. Second, we need to find out how/if we can resolve them. 

We can take a two-pronged approach to compatibility checking. First, it would be nice to have a way to check if there were any known incompatibilities. I mentioned an API endpoint related to updates above. We could also check the local dependencies definition (assuming one exists). Second, we could actually do a dry-run of the update and see if things explode. This will only detect huge breaks (things that the module can already partly mitigate via `*_exists()` checks as noted above). However, maybe we could go beyond that. Think really big and sophisticated. Maybe we could actually run a suite of tests that the module offers, and if those fail we know that there is an incompatibility.

Resolving the incompatibility could be as simple as checking for an update for the module, and offering to install it if one exists. Other than that, all we can do is tell the user that if they want to update they'll need to deactivate the module, or risk breaking their site.

Notes:
- Running dry-run updates with automated tests actually looks like the future to me. This is as close as you can humanly get to nothing-breaks updates. While it might take a lot of work to implement at first, I think it would be really worth it. We can start with simple dry-runs and work on the testing part as we go along. Related: WordPoints/wordpointsorg#35.

---

Problem 4: User updates a module which no longer supports the installed version of a plugin.

Why this is a problem: Things might break, causing concern, frustration, injury or even (the white screen of) death.

Minimum Solution: Show the user a message telling them that the installed version of the plugin is no longer supported and they need to update.

WWJD Solution: Automatically upgrade the plugin, or don't upgrade the module, depending on the user's intention.

Real Life Solution: When the user updates the module, give them the option to update the plugin as well.

Notes:
- In this scenario we assume that we already know about the incompatibility via the update API. However, dry-run checks as noted in problem 3 could still be useful.

**Comment by JDGrimes** [comment]
I think there are several things that we can glean from looking at the objections to this put forward by core devs:
- Modules shouldn't break when dependencies aren't installed or are the wrong version. I mean "break" as in WSOD. 
- This is done by using `*_exists()` calls as needed, but especially by not executing code that isn't attached to the appropriate hooks, mostly hooks _in the plugin the module is dependent on_.
- We're not talking about "framework"-type plugins that have no forward-facing features, that is a completely separate topic to come later.
- We're not trying to improve developer experience. Our focus is mostly user experience. We want to provide a bootstrap that developers can use to give an almost-or-maybe-even-really ideal experience to users of integrating modules.
- We aren't attempting to handle dependencies of things like PHP, MySQL, etc. Users don't know what those are and have no control. The requirements for these should always be assumed to be the same as for the supported WordPoints versions.
- Because plugins are visible things, we need to be transparent when handling them to avoid confusing users.

**Comment by JDGrimes** [comment]
### Dependency Conflicts

One of the blockers that people run up against is how to handle dependency conflicts between two different packages (in this case, modules).

**This problem already exists.** Any time you have installed multiple modules that integrate with a single plugin, you are already running the risk of dependency conflicts.

So:
- **This is not a new issue being introduced by us.** We are not exacerbating it either. Remember that we aren't trying to create a new kind of plugin that is _only_ a dependency. Rather, we are trying to improve user experience when two plugins integrate (that is, when a module integrates with another plugin).
- **This is actually problem 5 that we should try to solve.**

#### Problem 5: User installs to modules with conflicting dependencies

Why this is a problem: One or both of the modules may break, or cease to function fully.

Minimum Solution: Modules show a message telling the user that the dependent plugin is the wrong version. (This will probably just confuse a frustrate the user in many cases.)

WWJD Solution: The conflict is immediately resolved.

Real Life Solution: We need to know whether we can resolve the conflict, and if so how the user would want it to be done.

We can determine whether the conflict can be resolved by looking at each modules dependencies and determining whether a happy medium exists. We can also check whether there are updates available for any of the entities involved, and see if they will help resolve the conflict.

Notes:
- Most importantly, by solving the previous problems, we will actually prevent a module with a dependency conflict from ever being installed on the site. That doesn't actually solve the conflict itself though.
- We should be smart about this when plugins/modules are being installed/updated, so that we don't just tell the user not to install the module if the conflict could be resolved. We also don't want to tell them to update/install when it will cause a dependency conflict.

**Comment by JDGrimes** [comment]
Something that is important to keep in mind here, is that we can't assume anything about the plugin that is being integrated with. We have no control over any aspect of it, or the development practices.

So:
- We can't assume semantic versioning of dependencies.
- We can't assume that dependencies will keep any sort of backward compatibility.
- We can't assume that dependencies will be testable for issues.

In short, we can't assume compatibility just because "it's only a minor update".

**Comment by JDGrimes** [comment]
Prioritizing here, the first thing we need to do is establish a standard for specifying dependencies so that modules can begin doing this. We can then begin working on the implementation, and modules will already be ready when we actually release.

Let's create an RFC.

**Comment by JDGrimes** [comment]
> Let's create an RFC.

See #2. This currently only covers the formalized scope of the project. I'll open new tickets for things that need to be added.

**Comment by JDGrimes** [comment]
Another possible relationship type: `supersedes`/`superseded` for when one thing is replaced by or gets merged into another.

**Comment by JDGrimes** [comment]
I think that we should build the infrastructure for this in such a manner that it is an open standard. In other words, we shouldn't limit the types of dependencies based on just what we currently have. Each type should be handled by a separate handler, and if they handler for a certain type isn't present, we'll bail out and inform the user. So there will be one handler for plugins, another for modules, etc. The same is true of the relationship types, there should be one handler for required deps, and another for recommended, etc. That way new types could be added "on the fly", as it were.

**Comment by JDGrimes** [comment]
Another dependency type might be the WordPoints component.

**Comment by JDGrimes** [comment]
I think that maybe we should also make it possible to source the dependency information from multiple locations. So that even though we come up with a standard here, it will be easy to transition to the WordPress standard (if one ever exists) or support other possibilities in the future.

**Comment by JDGrimes** [comment]
Note that testing is one place where dependencies will prove useful. It tells us which dependencies need to be installed in order to test the module. So I think we need to keep in mind that we'll want to parse this during testing. Which since many of our tests run on PHP, shouldn't be a problem if we use HTML. Bash doesn't include build-in HTML parsing, but we probably don't generally need it at that level anyway, at least not as much.

**Comment by JDGrimes** [comment]
Maybe we should as a rule only have one module dependent on a plugin. There shouldn't be two modules that integrate with BuddyPress, for example. However, I'm not sure that that will always make sense. What if there is a module that provides another feature, but would also integrate nicely with certain aspects of BuddyPress? I suppose that instead of integrating with BuddyPress within that module, we would integrate that module within the BuddyPress module, if it was installed. I suppose that there could be cases where that wouldn't be ideal, because the BuddyPress module itself wouldn't really be needed and maybe the user didn't even want its other features. But generally speaking I think it would work will. Which would virtually eliminate the issue of dependency conflicts for us, though not if other devs are building lots of modules.

**Comment by JDGrimes** [comment]
We'll also want to utilize this information on WordPoints.org when providing the user information on module requirements.

Along similar lines, the use of JSON makes sense since we'd also use that in the updates API.
